storage/jellyfish_merkle/src/node_type/mod.rs

```diff
@@ -0,0 +1,505 @@
+                                panic!(
+                                    "Corrupted internal node: child_bitmap indicates \
+                                     the existence of a non-exist child at index {}",
+                                    only_child_index
+                                )
+                            })
+                            .version;
+                        Some(node_key.gen_child_node_key(only_child_version, only_child_index))
+                    },
+                    siblings,
+                );
+            }
+        }
+        unreachable!()
+    }
+}
+
+impl LeafNode {
+    /// Creates a new leaf node.
+    pub fn new(account_key: HashValue, blob: AccountStateBlob) -> Self {
+        let blob_hash = blob.hash();
+        Self {
+            account_key,
+            blob_hash,
+            blob,
+        }
+    }
+
+    /// Gets the account key, the hashed account address.
+    pub fn account_key(&self) -> HashValue {
+        self.account_key
+    }
+
+    /// Gets the hash of associated blob.
+    pub fn blob_hash(&self) -> HashValue {
+        self.blob_hash
+    }
+
+    /// Gets the associated blob itself.
+    pub fn blob(&self) -> &AccountStateBlob {
+        &self.blob
+    }
+}
+
+impl Node {
+    /// Creates the [`Internal`](Node::Internal) variant.
+    pub fn new_internal(children: HashMap<u8, Child>) -> Self {
+        Node::Internal(InternalNode::new(children))
+    }
+
+    /// Creates the [`Leaf`](Node::Leaf) variant.
+    pub fn new_leaf(account_key: HashValue, blob: AccountStateBlob) -> Self {
+        Node::Leaf(LeafNode::new(account_key, blob))
+    }
+
+    /// Returns `true` if the node is a leaf node.
+    pub fn is_leaf(&self) -> bool {
+        match self {
+            Node::Leaf(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Serializes to bytes for physical storage.
+    pub fn encode(&self) -> Result<Vec<u8>> {
+        let mut out = vec![];
+        match self {
+            Node::Internal(internal_node) => {
+                out.push(InternalNode::TAG);
+                out.extend(serialize(&internal_node)?);
+            }
+            Node::Leaf(leaf_node) => {
+                out.push(LeafNode::TAG);
+                out.extend(serialize(&leaf_node)?);
+            }
+        }
+        Ok(out)
+    }
+
+    /// Hashes are used to lookup the node in the database.
+    pub fn hash(&self) -> HashValue {
+        match self {
+            Node::Internal(internal_node) => internal_node.hash(),
+            Node::Leaf(leaf_node) => leaf_node.hash(),
+        }
+    }
+
+    /// Recovers from serialized bytes in physical storage.
+    pub fn decode(val: &[u8]) -> Result<Node> {
+        if val.is_empty() {
+            return Err(NodeDecodeError::EmptyInput.into());
+        }
+        let node_tag = val[0];
+        match node_tag {
+            InternalNode::TAG => Ok(Node::Internal(deserialize(&val[1..].to_vec())?)),
```

Comment:
looks like `.to_vec()` is not necessary?

Reply:
I bindly copied these lines. Will update.